test-notebooks/Dask_Cluster_Benchmark_notebook.ipynb

```diff
@@ -0,0 +1,166 @@
+    "\n",
+    "* Secure by Default: Cluster communication is automatically encrypted with TLS. All operations are authenticated with a configurable protocol, allowing you to use what makes sense for your organization.\n",
+    "\n",
+    "* Flexible: The gateway is designed to support multiple backends, and runs equally well in the cloud as on-premise. Natively supports Kubernetes, Hadoop/YARN, and HPC Job Queueing systems.\n",
+    "\n",
+    "* Robust to Failure: The gateway can be restarted or experience failover without losing existing clusters. Allows for seamless upgrades and restarts without disrupting users."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "13f24e39-bc62-4539-a15b-81c1639209e0",
+   "metadata": {},
+   "source": [
+    "<h3>Architecture Overview</h3>\n",
+    "Dask Gateway is divided into three separate components:\n",
+    "\n",
+    "Multiple active Dask Clusters (potentially more than one per user)\n",
+    "\n",
+    "A Proxy for proxying both the connection between the user’s client and their respective scheduler, and the Dask Web UI for each cluster\n",
+    "\n",
+    "A central Gateway that manages authentication and cluster startup/shutdown\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "430bc176-6595-4942-b78b-37922920abb9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# import dask, time, torch\n",
+    "from dask_gateway import Gateway\n",
+    "import time\n",
+    "import torch\n",
+    "gateway = Gateway( \"http://10.107.108.18\", auth=\"jupyterhub\", )\n",
+    "options = gateway.cluster_options()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c93e7c89-26f5-4b78-8eb2-96bc483b4c8c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create new cluster\n",
+    "cluster = gateway.new_cluster(options)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "98e43efd-c049-43f2-ab04-cc1048ff1b9c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# make the cluster scalable\n",
+    "cluster.adapt(minimum=2, maximum=10)\n",
+    "# showcase the gateway cluster information\n",
+    "cluster"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b8a2aeaa-0623-4d51-9849-31ddb52e6eed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import dask.array as da"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "619cf899-f116-4878-974c-6b9fe4222987",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "start_time = time.time()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "27bcabe2-08a6-405e-8780-96bf7422c96c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# cluster array \n",
+    "arr = da.random.random((1000, 1000), chunks=(1000,1000))\n",
+    "for _ in range(10000):\n",
+    "    arr += arr\n",
```

Comment:
you could move the start_time to be right before the for loop so you are only measuring the time it takes for arr += arr to run, and you could divide elapsed_time by 10000 (maybe make this an iterations variable) to get the average time.

Reply:
Same for the CPU array

Reply:
I was doing some benchmarking two weeks ago and I would run one iteration and then have the code wait for a second so the cluster could kill some of its workers just to test the worker create time and run time